rfd/0080-yubikey-piv.md

```diff
@@ -0,0 +1,272 @@
+auth_service:
+  ...
+  authentication:
+    ...
+    yubikey_login: 
+      required: true
+      pin_policy: always
+      touch_policy: always
+```
+
+Or with a dynamic Cluster Auth Preference object:
+
+```yaml
+kind: cluster_auth_preference
+version: v2
+metadata:
+  name: cluster-auth-preference
+spec:
+  yubikey_login: 
+    required: true
+    pin_policy: always
+    touch_policy: always
+```
+
+If `yubikey_login_required` is `false`, the policy fields will be ignored and attestation will be skipped.
+
+When `yubikey_login_required` is `true`, then the server will perform Attestation and verify the passed public key.
+
+If `pin_policy` or `touch_policy` is set, then the Auth server will check the [`Attestation`](https://pkg.go.dev/github.com/go-piv/piv-go@v1.9.0/piv#Attestation) object's `PINPolicy` or `TouchPolicy` fields. If the checked fields are at least as strict as the policies required, then the request will proceed.
+
+Examples:
+ - `pin_policy=once` + `PINPolicy=always` = success
+ - `touch_policy=always` + `TouchPolicy=always` = success
+ - `touch_policy=cached` + `TouchPolicy=never` = failure
+
+#### Role Options
+
+To make yubikey login required or optional on a per-role basis, you can set the same options in a Role.
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: role-name
+spec:
+  yubikey_login: 
+    required: true
+    pin_policy: once
+    touch_policy: cached
+```
+
+#### Combined yubikey requirements
+
+Each user's yubikey requirements will be determined by:
+ 1. Combining their role options to form the strictest setting
+ 2. Combining the strictest role options settings with the cluster auth preference, with role options taking precedence (if set).
+
+Note: A user's yubikey requirement settings cannot be easily determined before generating the yubikey private key and locking in its pin/touch policy settings, which occurs before logging in. Thus, it is the responsibility of the user to figure out their own pin/touch policy settings before logging in, or else they will get `access denied` errors.
+
+#### Examples:
+
+Example 1:
+ - All users must use yubikey login
+ - Users with the `admin` role must also use `pin=always` and `touch=always` (to secure adminstrative infrastructure)
+ - Users with the `superadmin` role are not required to use yubikey login (superadmin roles are inherently unsafe, might as well own it)
+```yaml
+kind: cluster_auth_preference
+version: v2
+metadata:
+  name: cluster-auth-preference
+spec:
+  yubikey_login: 
+    required: true
+---
+kind: role
+version: v5
+metadata:
+  name: admin
+spec:
+  yubikey_login: 
+    pin_policy: always
+    touch_policy: always
+---
+kind: role
+version: v5
+metadata:
+  name: superadmin
+spec:
+  yubikey_login: 
+    required: false
+```
+
+Example 2:
+ - Users with `yubikey-pin` role must use yubikey login with `pin=always` and `touch=cached|always`
+ - Users with `yubikey-touch` role must use yubikey login with `pin=once|always` and `touch=always`
```

Comment:
Where did `touch=always` came from?

Reply:
Ah typo, `yubikey-touch` should just have `touch_policy=always`